### PR TITLE
BMP-345 - handle error log

### PR DESCRIPTION
**Pull request by Danielecina**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks backward-compatibility?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any
| Updated README/docs?   | ✔️
| Added CHANGELOG entry?   | ✔️

## 📹 Video
erroronfield

### CODE REVIEW

File: src/utilities.js [hunk 1]
```diff
@@ -27,16 +27,18 @@ export function deepCopy (target) {
...
-      if (source[property] && isPlainObject(source[property])) {
-        if (!hasOwnProperty(destination, property)) destination[property] = {}
-        extend(destination[property], source[property])
-      } else if (Array.isArray(source[property])) {
-        destination[property] = deepCopy(source[property])
-      } else {
-        destination[property] = source[property]
-      }
-    })
+    if (source) {
```
Comment (Danielecina) [inline src/utilities.js]: in some cases `source` is` undefined` so it invokes a type error

File: src/validator.js [hunk 2]
```diff
@@ -337,12 +337,19 @@ export class Validator {
         const errors = []
         if (Array.isArray(schema.required)) {
           schema.required.forEach(e => {
+            if (this._checkType('string', value[e]) && value[e].length === 0) {
```
Comment (Danielecina) [inline src/validator.js:342]: check string cases when they are required and values change from stringsWithALength to “”

File: src/validator.js [hunk 3]
```diff
@@ -337,12 +337,19 @@ export class Validator {
...
+                message: this.translate('error_required', [e])
+              })
+            }
             if (typeof value[e] !== 'undefined') return
             const editor = this.jsoneditor.getEditor(`${path}.${e}`)
             /* Ignore required error if editor is of type "button" or "info" */
             if (editor && ['button', 'info'].includes(editor.schema.format || editor.schema.type)) return
             errors.push({
-              path,
+              path: `${path}.${e}`,
```
Comment (Danielecina) [inline src/validator.js]: this line allows to generate the errors inside the child fields

File: package.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "@json-editor/json-editor",
+  "name": "@mia-platform/json-editor",
```
Comment (dibericky) [inline package.json]: On `master` we keep `json-editor`.
Before releasing our version, we'll make a branch where we set `mia-platorm`

File: src/validator.js (hunk 3 above)
Comment (dibericky) [reply src/validator.js]: Are you sure that this is not `breaking-change`?

Comment (dibericky) [reply src/validator.js]: If it is, this should be based on an `option` with default value that keeps the current behaviour

File: src/themes/bootstrap4.js [hunk 5]
```diff
@@ -407,6 +407,7 @@ export class bootstrap4Theme extends AbstractTheme {
     }
 
     input.errmsg.textContent = text
+    return input
```
Comment (dibericky) [inline src/themes/bootstrap4.js]: In changes does not seem that your are using this method.
Where is it called?

File: package.json (hunk 4 above)
Comment (fredmaggiowski) [reply package.json]:
```suggestion
  "name": "@json-editor/json-editor",
```

File: src/themes/bootstrap4.js (hunk 5 above)
Comment (fredmaggiowski) [reply src/themes/bootstrap4.js]: This breaks the interface of the file, `add` functions does not return anything, also when you call `addInputError` you're already providing the reference to `input` so you don't need it to be returned


```suggestion
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,6 +3,7 @@
 - added option `use_default_values` used to specify if default values based on the "type" of the property have to be used
 - when `use_default_values` is false, number and integer fields have undefined value when input is empty
 - when `use_default_values` is false, string fields have undefined value at the beginning. After that user edit the field, empty input is threated as an empty string
+- changed position of errors logs down to each field
```
Comment (fredmaggiowski) [inline CHANGELOG.md]:
```suggestion
- changed position of errors logs below each field 
```

File: src/utilities.js (hunk 1 above)
Comment (fredmaggiowski) [reply src/utilities.js]:
```
if (!source) {
  return 
}

Object.keys()...
```

File: src/validator.js (hunk 3 above)
Comment (fredmaggiowski) [reply src/validator.js]: Let's make this optional and disabled by default

**Review by fredmaggiowski — CHANGES_REQUESTED** [review]
Let's make this feature optional with a documented option

File: src/defaults.js [hunk 7]
```diff
@@ -284,12 +284,15 @@ function translate (key, variables) {
   return string
 }
 
+export const UNDER_FIELD = 'under_field' // ['under_field', 'grouped']
```
Comment (Danielecina) [inline src/defaults.js]:
```suggestion
export const UNDER_FIELD = 'under_field'
```

**Review by davidebianchi — APPROVED** [review]

File: src/defaults.js [hunk 8]
```diff
@@ -284,12 +284,15 @@ function translate (key, variables) {
   return string
 }
 
+export const UNDER_FIELD = 'under_field'
```
Comment (fredmaggiowski) [inline src/defaults.js]: `below_field` would be a better value

File: README.md [hunk 9]
```diff
@@ -286,6 +286,11 @@ Here are all the available options:
     <td>If true default values based on the "type" of the property will be used</td>
     <td><code>true</code></td>
   </tr>
+  <tr>
+    <td>validation_error_placement</td>
+    <td>Change position where errors are shown</td>
```
Comment (fredmaggiowski) [inline README.md]: please document here also other available values

**Review by fredmaggiowski — CHANGES_REQUESTED** [review]

**Review by fredmaggiowski — APPROVED** [review]

**Pull request merged**